### DISCUSSION TITLE
How to read hashed password stored as a BLOB in SQLite

### DISCUSSION THREAD

**Discussion by zappfinger** [fastapi/fastapi#6963] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Description

I am using FastAPI with a SQLite database. The user table has name as a string and hashed_password as a BLOB in that database. In plain Python I can store and access the encrypted password like so: To store:

hashed_password = xor_strings(password.encode('cp1252'), key) #hashed_password is a byte string

to retrieve I use: decode('cp1252') on the query result. (Note that I am not using bcrypt)

In my schemas.py I have this Pydantic model:

```
class User(UserBase):
    id: int
    name: str
    hashed_password: str    # was bytes
```

But I get the following error:
```
File "/home/richard/.local/lib/python3.8/site-packages/fastapi/routing.py", line 126, in serialize_response
    raise ValidationError(errors, field.type_)
pydantic.error_wrappers.ValidationError: 1 validation error for User
response -> hashed_password
  'utf-8' codec can't decode byte 0xe2 in position 1: invalid continuation byte (type=value_error.unicodedecode)
```

How can I force the codec to use 'cp1252' and not 'utf-8'?
Is my datatype (str) correct?

### Environment

* OS: Linux Mint
* FastAPI Version 0.63.0

* Python version: 3.8

**Comment by Mause** [comment]
Just to clarify, you are returning a "hashed" password from an endpoint?

It's worth noting that json doesn't support raw bytes, if you want to return bytes you should consider base 64 encoding them

**Comment by zappfinger** [comment]
> Just to clarify, you are returning a "hashed" password from an endpoint?
> 
> It's worth noting that json doesn't support raw bytes, if you want to return bytes you should consider base 64 encoding them
What do you mean by endpoint? It comes from a database record. I then want to compare it with a password the user enters to login (after the same encryption).
How and where should I encode base 64? Note that the plain Python example has an encoding that works

**Comment by Mause** [comment]
If you just want to store raw bytes in a model, then use the `bytes` type?

Also, I hope you realise that that isn't really encryption at all, it's hashing, and weak hashing at that, and shouldn't be used in a real world application - scrypt or bcrypt should be used instead

**Comment by zappfinger** [comment]
> If you just want to store raw bytes in a model, then use the `bytes` type?
I also tried the bytes type, it gave another error:
```
  File "/home/richard/.local/lib/python3.8/site-packages/fastapi/encoders.py", line 90, in jsonable_encoder
    encoded_value = jsonable_encoder(
  File "/home/richard/.local/lib/python3.8/site-packages/fastapi/encoders.py", line 127, in jsonable_encoder
    return ENCODERS_BY_TYPE[type(obj)](obj)
  File "pydantic/json.py", line 22, in pydantic.json.lambda
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe2 in position 1: invalid continuation byte
```
It might not be true encryption, but I created my own hashing routine using XOR

**Comment by Mause** [comment, marked as answer]
You will only get that error if you are trying to return a model containing bytes from a route/endpoint, which again, isn't supported in json. If you wish to return bytes, you should first encode them as base64

But I would advise you to reconsider returning a user's "encrypted" password from a route/endpoint, as that would be a major security flaw

**Comment by zappfinger** [comment]
Ok thanks, I'll go that path...

**Comment by tiangolo** [comment]
Thanks for the help here @Mause ! 👏  🙇 

Thanks for reporting back and closing the issue @zappfinger  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
